File: passenger-service/src/main/java/com/example/passenger_service/config/KafkaConfig.java
package com.example.passenger_service.config;

import com.example.passenger_service.dto.PassengerRatingEvent;
import org.apache.kafka.clients.consumer.ConsumerConfig;
import org.apache.kafka.common.serialization.StringDeserializer;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.kafka.config.ConcurrentKafkaListenerContainerFactory;
import org.springframework.kafka.core.ConsumerFactory;
import org.springframework.kafka.core.DefaultKafkaConsumerFactory;
import org.springframework.kafka.support.serializer.ErrorHandlingDeserializer;
import org.springframework.kafka.support.serializer.JsonDeserializer;

import java.util.HashMap;
import java.util.Map;

@Configuration
public class KafkaConfig {

    @Bean
    public ConsumerFactory<String, PassengerRatingEvent> passengerRatingConsumerFactory() {
        Map<String, Object> props = new HashMap<>();
        props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:29092");
        props.put(ConsumerConfig.GROUP_ID_CONFIG, "passenger-rating-group");
        props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");

        props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, ErrorHandlingDeserializer.class);
        props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, ErrorHandlingDeserializer.class);

        props.put(ErrorHandlingDeserializer.KEY_DESERIALIZER_CLASS, StringDeserializer.class);
        props.put(ErrorHandlingDeserializer.VALUE_DESERIALIZER_CLASS, JsonDeserializer.class);

        props.put(JsonDeserializer.VALUE_DEFAULT_TYPE, PassengerRatingEvent.class.getName());
        props.put(JsonDeserializer.TRUSTED_PACKAGES, "com.example.passenger_service.dto");

        return new DefaultKafkaConsumerFactory<>(props);
    }

    @Bean
    public ConcurrentKafkaListenerContainerFactory<String, PassengerRatingEvent>
    passengerRatingListenerContainerFactory() {

        ConcurrentKafkaListenerContainerFactory<String, PassengerRatingEvent> factory =
                new ConcurrentKafkaListenerContainerFactory<>();
        factory.setConsumerFactory(passengerRatingConsumerFactory());
        return factory;
    }
}

File: driver-service/src/main/kotlin/com/example/driver_service/config/KafkaConfig.kt
package com.example.driver_service.config

import com.example.driver_service.dto.DriverNotification
import com.example.driver_service.dto.DriverRatingEvent
import org.apache.kafka.clients.consumer.ConsumerConfig
import org.apache.kafka.common.serialization.StringDeserializer
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.kafka.config.ConcurrentKafkaListenerContainerFactory
import org.springframework.kafka.core.ConsumerFactory
import org.springframework.kafka.core.DefaultKafkaConsumerFactory
import org.springframework.kafka.support.serializer.JsonDeserializer

@Configuration
class KafkaConfig {

    private fun commonConsumerProps(groupId: String): Map<String, Any> = mapOf(
        ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG to "localhost:29092",
        ConsumerConfig.GROUP_ID_CONFIG to groupId,
        ConsumerConfig.AUTO_OFFSET_RESET_CONFIG to "earliest",
        ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG to false
    )

    @Bean
    fun driverNotificationConsumerFactory(): ConsumerFactory<String, DriverNotification> {
        return DefaultKafkaConsumerFactory(
            commonConsumerProps("driver-service-group"),
            StringDeserializer(),
            JsonDeserializer(DriverNotification::class.java).apply {
                addTrustedPackages("com.example.driver_service.dto")
            }
        )
    }

    @Bean
    fun driverNotificationListenerContainerFactory(): ConcurrentKafkaListenerContainerFactory<String, DriverNotification> {
        return ConcurrentKafkaListenerContainerFactory<String, DriverNotification>().apply {
            consumerFactory = driverNotificationConsumerFactory()
        }
    }

    @Bean
    fun driverRatingConsumerFactory(): ConsumerFactory<String, DriverRatingEvent> {
        return DefaultKafkaConsumerFactory(
            commonConsumerProps("driver-rating-group"),
            StringDeserializer(),
            JsonDeserializer(DriverRatingEvent::class.java).apply {
                addTrustedPackages("com.example.driver_service.dto")
            }
        )
    }

    @Bean
    fun driverRatingListenerContainerFactory(): ConcurrentKafkaListenerContainerFactory<String, DriverRatingEvent> {
        return ConcurrentKafkaListenerContainerFactory<String, DriverRatingEvent>().apply {
            consumerFactory = driverRatingConsumerFactory()
        }
    }
}